src/index.ts

```diff
@@ -132,15 +155,96 @@ function addSegmentFromDirectionsResponse(previousLngLat: LngLat, e: MapMouseEve
     });
 }
 
-function addSegmentFromStraightLine(previousLngLat: LngLat, e: MapMouseEvent): void {
-  const newSegment = nextSegmentService.segmentFromStraightLine(previousLngLat, e.lngLat);
+function addSegmentFromStraightLine(previousLngLat: LngLat, lngLat: LngLat, animate = true): void {
+  const newSegment = nextSegmentService.segmentFromStraightLine(previousLngLat, lngLat);
 
-  animationService.animateSegment(newSegment);
-  const marker = addMarker(e.lngLat, false);
+  if (animate) animationService.animateSegment(newSegment);
+  const marker = addMarker(lngLat, false);
   currentRun.addSegment(newSegment, marker);
   updateLengthElement();
 }
 
+function runToJson(run: CurrentRun): string {
+  if (run === undefined) return "{}";
+  let runJSON: {[name:string]: any} = {
+    start: {
+      lng: run.start.lngLat.lng,
+      lat: run.start.lngLat.lat,
+    },
+    distance: run.distance,
+    segments: [],
+    followRoads
+  };
+  for (let i in run.segments) {
+    runJSON.segments.push({
+      lng: run.segments[i].lngLat.lng,
+      lat: run.segments[i].lngLat.lat,
+      followsRoads: run.segments[i].followsRoads
+    });
+  }
+  return JSON.stringify(runJSON);
+}
+
+function jsonToRun(json: string): boolean {
+  try { 
+    let runJSON = JSON.parse(json);
+    let lngLat = new LngLat(runJSON.start.lng, runJSON.start.lat);
+    let start = new RunStart(lngLat);
+    start.setMarker(addMarker(lngLat, true));
+    let newRun = new CurrentRun(start);
+    let prev = lngLat;
+    for (let i = 0; i < runJSON.segments.length; i++) {
+      let lngLat = new LngLat(runJSON.segments[i].lng, runJSON.segments[i].lat);
+      if (runJSON.segments[i].followsRoads) {
+        addSegmentFromDirectionsResponse(prev, lngLat, false);
+      } else {
+        addSegmentFromStraightLine(prev, lngLat, false);
+      }
+      prev = lngLat;
+    }
+    clearRun();
+    currentRun = newRun;
+    return true;
+  }
+  catch (err) {
+    console.log(err);
+    return false;
+  }
+  finally {
+    setTimeout(() => animationService.readdRunToMap(currentRun), 150);
+  }
+}
+
+function downloadRun(): void {
+  let run = runToJson(currentRun);
+  let file = new Blob([run], {
+    type: "application/json"
+  });
+  let url = URL.createObjectURL(file);
+  let link = document.createElement("a");
+  link.href = url;
+  link.download = "currentRun.runmap";
+  link.click();
+}
+
+function showUploadForm(): void {
+  closeMenu(false);
+  uploadContainer.classList.add("showing-form");
+  uploadContainer.setAttribute('aria-hidden', 'false');
+  runInput.value = "";
```

Comment:
with the value being cleared, should the inner text also be cleared just in case it's still set to the previous filename? I noticed if I loaded a run, then re-opened the load modal it still seemed to show the loaded filename - minor, but might be worth fixing now

Reply:
I'll do this